packages/boot/test/bootstrapTests/test-net-ibc-upgrade.ts

```diff
@@ -0,0 +1,215 @@
+  for (const [name, path] of entries(asset)) {
+    const bundle = await bundleCache.load(path, name);
+    const installation = await EV(zoe).install(bundle);
+    t.truthy(installation);
+    assign(t.context.installation, { [name]: installation });
+  }
+});
+
+const upgradeVats = async (t, EV, vatsToUpgrade) => {
+  const vatStore = await EV.vat('bootstrap').consumeItem('vatStore');
+  const vatUpgradeInfo =
+    await EV.vat('bootstrap').consumeItem('vatUpgradeInfo');
+  const vatAdminSvc = await EV.vat('bootstrap').consumeItem('vatAdminSvc');
+  for (const vatName of vatsToUpgrade) {
+    const { bundleID } = await EV(vatUpgradeInfo).get(vatName);
+    const bcap = await EV(vatAdminSvc).getBundleCap(bundleID);
+    const { adminNode } = await EV(vatStore).get(vatName);
+    const result = await EV(adminNode).upgrade(bcap);
+    t.log(vatName, result);
+    t.true(result.incarnationNumber > 0);
+  }
+};
+
+test.serial('upgrade at many points in network API flow', async t => {
+  const { installation } = t.context;
+  const { EV } = t.context.runUtils;
+  const networkVat = await EV.vat('bootstrap').consumeItem('networkVat');
+  const zoe: ZoeService = await EV.vat('bootstrap').consumeItem('zoe');
+
+  const flow = entries({
+    startServer: async label => {
+      const started = await EV(zoe).startInstance(
+        installation.ibcServerMock,
+        {},
+        {},
+        { address: '/ibc-port/', networkVat },
+      );
+      t.truthy(started.creatorFacet, `${label} ibcServerMock`);
+      return [label, { server: started.creatorFacet }];
+    },
+    startListening: async ([label, opts]) => {
+      await EV.sendOnly(opts.server).listen();
+      await EV.sendOnly(opts.server).dequeue('onListen');
+      return [label, opts];
+    },
+    startClient: async ([label, opts]) => {
+      const started = await EV(zoe).startInstance(
+        installation.ibcClientMock,
+        {},
+        {},
+        { address: '/ibc-port/', networkVat },
+      );
+      t.truthy(started.creatorFacet, `${label} ibcClientMock`);
+      return [label, { ...opts, client: started.creatorFacet }];
+    },
+    getAddresses: async ([label, opts]) => {
+      const serverAddress = await EV(opts.server).getLocalAddress();
+      const clientAddress = await EV(opts.client).getLocalAddress();
+      t.log(`${label} server ${serverAddress} client ${clientAddress}`);
+      return [label, { ...opts, serverAddress }];
+    },
+    startConnecting: async ([label, opts]) => {
+      await EV.sendOnly(opts.client).connect(opts.serverAddress);
+      await EV.sendOnly(opts.server).dequeue('onAccept');
+      await EV.sendOnly(opts.server).dequeue('onOpen');
+      return [label, opts];
+    },
+    sendPacket: async ([label, opts]) => {
+      await EV.sendOnly(opts.client).send(label);
+      return [label, opts];
+    },
+    respond: async ([label, opts]) => {
+      await EV.sendOnly(opts.server).dequeue('onReceive');
+      return [label, opts];
+    },
```

Comment:
Formally capturing something I had previously only said in more ephemeral channels: each `.dequeue(…)` should have its own step, to test for interruption of the affected promise.
```suggestion
    requestListening: async ([label, opts]) => {
      await EV.sendOnly(opts.server).listen();
      return [label, opts];
    },
    startListening: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onListen');
      return [label, opts];
    },
    startClient: async ([label, opts]) => {
      const started = await EV(zoe).startInstance(
        installation.ibcClientMock,
        {},
        {},
        { address: '/ibc-port/', networkVat },
      );
      t.truthy(started.creatorFacet, `${label} ibcClientMock`);
      return [label, { ...opts, client: started.creatorFacet }];
    },
    getAddresses: async ([label, opts]) => {
      const serverAddress = await EV(opts.server).getLocalAddress();
      const clientAddress = await EV(opts.client).getLocalAddress();
      t.log(`${label} server ${serverAddress} client ${clientAddress}`);
      return [label, { ...opts, serverAddress }];
    },
    requestConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.client).connect(opts.serverAddress);
      return [label, opts];
    },
    acceptConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onAccept');
      return [label, opts];
    },
    openConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onOpen');
      return [label, opts];
    },
    sendPacket: async ([label, opts]) => {
      await EV.sendOnly(opts.client).send(label);
      return [label, opts];
    },
    respond: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onReceive');
      return [label, opts];
    },
```